File: macbook/bettertouchtool/install_performance_audio_menu.js
// Uses BTT's API only. A Tools button opens an independent vertical floating menu.
ObjC.import("Foundation");

function readJSON(path) {
    const text = $.NSString.stringWithContentsOfFileEncodingError(path, $.NSUTF8StringEncoding, null);
    if (!text) throw new Error("Cannot read " + path);
    return JSON.parse(ObjC.unwrap(text));
}
function isEnabled(item) {
    return ["BTTEnabled", "BTTEnabled2"].every(key => item[key] === undefined || Number(item[key]) === 1);
}
function isHiddenMenuItem(item) {
    // Disabled items can disappear from a parent query. Other BTT exports omit
    // enabled fields even when disabled; explicit visibility provides readback.
    if (!item || !isEnabled(item)) return true;
    const config = item.BTTMenuConfig || {};
    return config.BTTMenuItemVisibleWhileActive !== undefined &&
        config.BTTMenuItemVisibleWhileInactive !== undefined &&
        Number(config.BTTMenuItemVisibleWhileActive) === 0 &&
        Number(config.BTTMenuItemVisibleWhileInactive) === 0;
}
function summary(item) {
    const config = item.BTTMenuConfig || {};
    return {uuid: item.BTTUUID, name: item.BTTMenuName || item.BTTTriggerName,
        type: item.BTTTriggerType, enabled: isEnabled(item),
        layout: config.BTTMenuLayoutDirection};
}
function actionData(action) {
    const data = action.BTTAdditionalActionData || {};
    return typeof data === "string" ? JSON.parse(data) : data;
}
function matchesItem(actual, expected) {
    if (!actual || !isEnabled(actual) || actual.BTTUUID !== expected.BTTUUID ||
        Number(actual.BTTTriggerType) !== Number(expected.BTTTriggerType)) return false;
    const actions = actual.BTTMenuItemActions || [];
    const desired = expected.BTTMenuItemActions || [];
    return actions.length === desired.length && desired.every((wanted, index) => {
        const action = actions[index];
        const type = Number(wanted.BTTPredefinedActionType);
        if (Number(action.BTTPredefinedActionType) !== type || !isEnabled(action)) return false;
        if (type === 386) return actionData(action).BTTMenuActionMenuID === actionData(wanted).BTTMenuActionMenuID;
        return type === 206 && action.BTTShellTaskActionScript === wanted.BTTShellTaskActionScript &&
            action.BTTShellTaskActionConfig === wanted.BTTShellTaskActionConfig;
    });
}
function matchesLayout(item, layout) {
    const config = item.BTTMenuConfig || {};
    return Object.keys(layout).every(key => Number(config[key] === undefined ? 0 : config[key]) === layout[key]);
}

function run(argv) {
    const definition = readJSON(argv[0]);
    const dropdown = definition.dropdown;
    const launcher = definition.launcher;
    const btt = Application("/Applications/BetterTouchTool.app");
    function get(query) {
        const items = JSON.parse(btt.get_triggers(query));
        if (!Array.isArray(items)) throw new Error("Unexpected BTT response; refusing changes.");
        return items;
    }
    function childItems(id) { return get({trigger_parent_uuid: id}); }
    function one(items, id) {
        const found = items.filter(item => item.BTTUUID === id);
        if (found.length > 1) throw new Error("Duplicate UUID returned: " + id);
        return found[0];
    }
    const menus = get({trigger_id: 767});
    const parent = one(menus, definition.parent);
    if (!parent || Number(parent.BTTTriggerType) !== 767 || !isEnabled(parent)) {
        throw new Error("Expected enabled Media menu; no changes made. " + JSON.stringify(menus.map(summary)));
    }
    const children = childItems(parent.BTTUUID);
    const toolsMatches = children.filter(item => item.BTTUUID === definition.tools_uuid || item.BTTMenuName === "Tools");
    const tools = toolsMatches.length === 1 ? toolsMatches[0] : null;
    const toolsItems = tools ? childItems(tools.BTTUUID) : [];
    const legacy = one(toolsItems, definition.legacy_uuid);
    const existingLauncher = one(toolsItems, launcher.BTTUUID);
    const existingDropdown = one(menus, dropdown.BTTUUID);
    const dropdownItems = existingDropdown ? childItems(dropdown.BTTUUID) : [];
    if (argv[2] === "inspect") {
        console.log(JSON.stringify({parent: summary(parent), children: children.map(summary),
            tools: tools && summary(tools), toolsItems: toolsItems.map(summary),
            dropdown: existingDropdown && summary(existingDropdown),
            dropdownItems: dropdownItems.map(summary)}, null, 2));
        return;
    }
    if (!tools || Number(tools.BTTTriggerType) !== 774 || !isEnabled(tools)) {
        throw new Error("Expected exactly one enabled Tools submenu; no changes made. Media children: " + JSON.stringify(children.map(summary)));
    }
    if ((legacy && Number(legacy.BTTTriggerType) !== 774) ||
        (existingLauncher && Number(existingLauncher.BTTTriggerType) !== 773) ||
        (existingDropdown && Number(existingDropdown.BTTTriggerType) !== 767) ||
        toolsItems.some(item => item.BTTMenuName === "Performance Audio" &&
            item.BTTUUID !== definition.legacy_uuid && item.BTTUUID !== launcher.BTTUUID)) {
        throw new Error("Conflicting Performance Audio item; refusing to replace an unknown item.");
    }
    const legacyItems = legacy ? childItems(legacy.BTTUUID) : [];
    const snapshot = JSON.stringify({parent, children, toolsItems, legacyItems,
        dropdown: existingDropdown || null, dropdownItems}, null, 2);
    const backup = argv[1] + "/btt-before-performance-audio-" + Date.now() + ".json";
    if (!$(snapshot).writeToFileAtomicallyEncodingError(backup, true, $.NSUTF8StringEncoding, null) ||
        !$.NSFileManager.defaultManager.setAttributesOfItemAtPathError(
            $({NSFilePosixPermissions: 384}), backup, null)) {
        throw new Error("Could not save a private backup; no changes made.");
    }
    function upsertItem(wanted, actual, parentID) {
        wanted.BTTTriggerParentUUID = parentID;
        if (!actual) {
            btt.add_new_trigger(JSON.stringify(wanted), {parent_uuid: parentID});
        } else if (!matchesItem(actual, wanted)) {
            const replacement = Object.assign({}, actual, wanted);
            if (actual.BTTOrder === undefined) delete replacement.BTTOrder;
            else replacement.BTTOrder = actual.BTTOrder;
            btt.update_trigger(wanted.BTTUUID, {json: JSON.stringify(replacement)});
        }
    }
    if (!existingDropdown) {
        // No parent_uuid: this is a genuine top-level menu with its own layout.
        btt.add_new_trigger(JSON.stringify(dropdown));
    } else {
        if (!isEnabled(existingDropdown) || !matchesLayout(existingDropdown, definition.dropdown_layout)) {
            btt.update_trigger(dropdown.BTTUUID, {json: JSON.stringify({BTTEnabled: 1,
                BTTMenuConfig: Object.assign({}, existingDropdown.BTTMenuConfig || {}, definition.dropdown_layout)})});
        }
        for (const wanted of dropdown.BTTMenuItems) {
            upsertItem(wanted, one(dropdownItems, wanted.BTTUUID), dropdown.BTTUUID);
        }
    }
    const savedDropdown = one(get({trigger_id: 767}), dropdown.BTTUUID);
    const savedItems = childItems(dropdown.BTTUUID);
    if (!savedDropdown || Number(savedDropdown.BTTTriggerType) !== 767 || !isEnabled(savedDropdown) ||
        !matchesLayout(savedDropdown, definition.dropdown_layout) ||
        !dropdown.BTTMenuItems.every(item => matchesItem(one(savedItems, item.BTTUUID), item))) {
        throw new Error("Dropdown verification failed; old submenu remains available. Backup: " + backup);
    }
    if (!existingLauncher && legacy && legacy.BTTOrder !== undefined) launcher.BTTOrder = legacy.BTTOrder;
    upsertItem(launcher, existingLauncher, tools.BTTUUID);
    if (!matchesItem(one(childItems(tools.BTTUUID), launcher.BTTUUID), launcher)) {
        throw new Error("Launcher verification failed; old submenu remains available. Backup: " + backup);
    }
    // Retire the old horizontal submenu only after its replacement is verified.
    // Keep it and all its descendants intact for recovery; never delete them.
    if (legacy && !isHiddenMenuItem(legacy)) {
        btt.update_trigger(legacy.BTTUUID, {json: JSON.stringify({
            BTTEnabled: 0, BTTEnabled2: 0,
            BTTMenuConfig: Object.assign({}, legacy.BTTMenuConfig || {}, {
                BTTMenuItemVisibleWhileActive: 0, BTTMenuItemVisibleWhileInactive: 0,
            }),
        })});
        const retired = one(childItems(tools.BTTUUID), legacy.BTTUUID);
        if (!isHiddenMenuItem(retired)) throw new Error("Could not hide the old submenu. Backup: " + backup);
    }
    console.log("Verified configuration: Media > Tools > Performance Audio opens a separate vertical dropdown. " +
        "Close and reopen Tools to use the new button. Backup: " + backup);
}

File: macbook/bettertouchtool/install_performance_audio_menu.py
#!/usr/bin/env python3
"""Add a Tools button that opens an independent vertical Performance Audio menu."""

import argparse
import copy
import json
import os
from pathlib import Path
import subprocess
import sys

from btt_touchbar_folder_to_floating_submenu import rtf_label
from build_performance_audio_menu import build, uid
from tools_menu_style import tools_button_config


ROOT = Path(__file__).resolve().parents[1]
MAIN_MENU = "D9B0ED12-C4BE-4E74-B0DA-0CC3BE092289"
# uuid5(install_strip_metadata_menu.NAMESPACE, "tools"); shared installer identity.
TOOLS_MENU = "FC3F8235-F102-55C4-8432-B6ADFB0D9992"


def definition(parent):
    dropdown = copy.deepcopy(build()["BTTPresetContent"][0]["BTTTriggers"][0])
    dropdown_id = uid("tools-dropdown")
    launcher_id = uid("tools-dropdown-launcher")
    dropdown.update(BTTUUID=dropdown_id, BTTMenuName="Performance Audio Tools Presets",
                    BTTTriggerName="Floating Menu: Performance Audio Tools Presets",
                    BTTAppBundleIdentifier="BT.G")
    # New IDs leave both the standalone pack and the old nested submenu intact.
    for item in dropdown["BTTMenuItems"]:
        item["BTTUUID"] = uid("tools-dropdown-" + item["BTTUUID"])
        item["BTTTriggerParentUUID"] = dropdown_id
        item["BTTMenuConfig"]["BTTMenuAttributedText"] = rtf_label(
            item["BTTMenuName"]).replace("\\fs50", "\\fs30")
        for action in item["BTTMenuItemActions"]:
            action["BTTUUID"] = uid("tools-dropdown-" + action["BTTUUID"])
            action["BTTTriggerParentUUID"] = item["BTTUUID"]
    config = tools_button_config("performance-audio-tools-launcher", "Performance Audio")
    layout = {
        "BTTMenuLayoutDirection": 6,  # verticalOneColumn
        "BTTMenuSizingBehavior": 1,
        "BTTMenuFrameWidth": 350,
        "BTTMenuFrameHeight": 44 * len(dropdown["BTTMenuItems"]) + 16,
        "BTTMenuHorizontalAlignment": 0,
        "BTTMenuVerticalAlignment": 0,
        "BTTMenuHorizontalSpacing": 0,
        "BTTMenuVerticalSpacing": 4,
        "BTTMenuPositioningType": 1,
        "BTTMenuPositionRelativeTo": 7,
        "BTTMenuVisibility": 1,
        "BTTMenuCloseAfterAction": 1,
        "BTTMenuKeepCached": 0,
    }
    dropdown["BTTMenuConfig"].update(layout)
    dropdown["BTTMenuConfig"]["BTTMenuElementIdentifier"] = "performance-audio-tools-dropdown"
    return {"parent": parent, "tools_uuid": TOOLS_MENU,
            "legacy_uuid": uid("tools-submenu"), "dropdown_layout": layout,
            "dropdown": dropdown, "launcher": {
        "BTTUUID": launcher_id, "BTTTriggerType": 773,
        "BTTTriggerTypeDescriptionReadOnly": "Standard Item",
        "BTTTriggerClass": "BTTTriggerTypeFloatingMenu", "BTTEnabled": 1,
        "BTTMenuName": "Performance Audio", "BTTMenuConfig": config,
        "BTTMenuItemActions": [{
            "BTTUUID": uid("tools-dropdown-show-action"), "BTTTriggerParentUUID": launcher_id,
            "BTTTriggerClass": "BTTTriggerTypeFloatingMenu", "BTTEnabled": 1, "BTTOrder": 0,
            "BTTPredefinedActionType": 386, "BTTPredefinedActionName": "Show Floating Menu",
            "BTTAdditionalActionData": {"BTTMenuActionMenuID": dropdown_id,
                                        "BTTMenuActionMenuName": dropdown["BTTMenuName"],
                                        "BTTMenuActionRestorePosition": False},
        }],
    }}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    mode = parser.add_mutually_exclusive_group()
    mode.add_argument("--apply", action="store_true", help="back up, install, and verify using BTT's API")
    mode.add_argument("--inspect", action="store_true", help="inspect Media > Tools without changing BTT")
    parser.add_argument("--parent-uuid", default=MAIN_MENU)
    args = parser.parse_args()
    os.umask(0o077)  # Live menu backups can contain private action payloads.
    build_dir = ROOT / "build"
    build_dir.mkdir(exist_ok=True)
    (build_dir / "performance-audio-jobs").mkdir(exist_ok=True)
    payload = build_dir / "Performance Audio Tools.json"
    payload.write_text(json.dumps(definition(args.parent_uuid), indent=2) + "\n")
    if not args.apply and not args.inspect:
        print(f"Generated {payload}; use --apply to install into Media > Tools.")
        return
    result = subprocess.run([
        "/usr/bin/osascript", "-l", "JavaScript",
        str(Path(__file__).with_suffix(".js")), str(payload), str(build_dir),
        "inspect" if args.inspect else "apply"])
    if result.returncode:
        print("BTT command failed; see the specific error above. Do not use sudo.", file=sys.stderr)
        raise SystemExit(result.returncode)


if __name__ == "__main__":
    main()
